src/os_common/os_alloc.c

```diff
@@ -114,11 +130,16 @@ __realloc_func(WT_SESSION_IMPL *session, size_t *bytes_allocated_ret, size_t byt
      * Sometimes we're allocating memory and we don't care about the
      * final length -- bytes_allocated_ret may be NULL.
      */
-    p = *(void **)retp;
     bytes_allocated = (bytes_allocated_ret == NULL) ? 0 : *bytes_allocated_ret;
-    WT_ASSERT(session,
-      (p == NULL && bytes_allocated == 0) ||
-        (p != NULL && (bytes_allocated_ret == NULL || bytes_allocated != 0)));
+    if (realloc_malloc)
+        WT_ASSERT(session,
+          (tmpp == NULL && bytes_allocated == 0) ||
+            (tmpp != NULL && (bytes_allocated_ret == NULL || bytes_allocated != 0)));
+    else
+        WT_ASSERT(session,
+          (p == NULL && bytes_allocated == 0) ||
+            (p != NULL && (bytes_allocated_ret == NULL || bytes_allocated != 0)));
+
```

Comment:
This is correct code, but I think it could be shorter.  Consider removing the conditional assignment of `p` at the top of the function, and always make `p = *(void **)retp;` (as it was before).  By doing that, we can leave the original ASSERT in place here.  And by doing that, we wouldn't need the `realloc_malloc` variable (since we're only checking it once), we can check the connection flag at the right spot.

Metagoal: fewer diffs in changes/smaller code means easier reviews, easier to reason about, bugs less likely.

Reply:
Looks fixed to me.